text/0083-rawkv-cross-cluster-replication.md

```diff
@@ -0,0 +1,185 @@
+
+#### High Latency between Clusters
+
+Customers require that [RPO](https://en.wikipedia.org/wiki/Disaster_recovery#Recovery_Point_Objective) should be no more than tens of seconds. As *RPO* is consist of duration in TiKV, TiKV-CDC *(see below)*, and network latency, and the first two can be improved by code optimization and scale-out, we should be careful to deal with the last.
+
+A typical ping latency of cross city network will be more than 30 milliseconds, which is much higher than IDC internal network. We should use appropriate concurrency and batch size to share latency to each data entry, at the same time limit transmission rate to avoid package loss.
+
+In this proposal, we provide a TiKV sink with configurable concurrency, batch size, and backoff parameters, to be adapted to different network environments.
+
+## Detailed Design
+
+### 1. Utilize TiKV-BR & TiKV-CDC as Data Replication Component
+
+**TiKV-BR**, a fork of [BR](https://github.com/pingcap/tidb/tree/master/br), is the tool for backup and restoration of TiKV. We utilize **TIKV-BR** to accomplish data initialization of recovery cluster.
+
+**TiKV-CDC**, a fork of [TiCDC](https://github.com/pingcap/tiflow), is the tool for replicating incremental data of TiKV, which gets change data from TiKV's observer component, with high performance, high available, and scalable.
+
+Besides, **TiKV-CDC** can also provide the ability to connect to other components, e.g. Message Queues, which will help TiKV integrate with customers' data systems, and further extend the applicable scenarios of TiKV.
+
+![rawkv-cdc](../media/rawkv-cdc.png)
+
+### 2. Add Timestamp to Data
+
+As a kind of [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture), timestamp (or version) is necessary to indicate which and when data is changed.
+
+#### 2.1 Requirement
+
+Among requests of a key, the timestamp must be monotonic with the sequence of data flushed to disk in TiKV.
+
+In general, if request `a` ["happened before"](https://en.wikipedia.org/wiki/Happened-before) `b`, then `Timestamp(a) < Timestamp(b)`. As to cross cluster replication, we provide [Causal Consistency](https://en.wikipedia.org/wiki/Causal_consistency) by keeping order of the timestamp the same as sequence of data flush to disk in TiKV. Downstream systems apply data according to the timestamp order, will always lead to the consistent result.
+
+At the same time, as RawKV doesn't provide cross-rows transaction and snapshot isolation, we treat requests of different keys as concurrent, to reduce complexity and improve performance.
+
+#### 2.2 Timestamp Generation
+
+Timestamp is generated by TiKV internally, to get a better overall performance and client compatibility.
+
+We use [HLC](https://cse.buffalo.edu/tech-reports/2014-04.pdf) method to generate timestamp, which not only is easy to use (by being closed to real world time), but also can capture causality.
```

Comment:
could you add move info about how many bits are physical and logical timestamp occupies?

Reply:
OK. Done, PTAL~